### PR TITLE
Fix client blocking

### PR DESCRIPTION
**Pull request by nickfh7**
During rpc client calls, the mutex on each client hashmap blocks all other client calls to the same service until the entire call has finished. Instead, client caching has been reworked only lock the hashmap until a clone of the client can be made. This gives good performance with minimal lock time.

### CODE REVIEW

**Comment by nickfh7** [comment]
Quick note: everything up to "Fix config address" (e0a3bc7) has already been merged

**Comment by matt9j** [comment]
The change here looks good to me! I would suggest rebasing this branch before merging, since it looks like commit [f61a003](https://github.com/uw-ictd/dAuth/pull/47/commits/f61a0036b3a0001d645fd4288f6c34fdc07da3fe) and [d37adc3](https://github.com/uw-ictd/dAuth/pull/47/commits/d37adc34b06ff6d6f578ea07d6ea5d69726a9554) are the same, and the following two commits seem like a spurious merge and a fixup.

**Review by matt9j — APPROVED** [review]
lgtm pending resolution of the merge weirdness.

**Pull request merged**
